fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexScrubbing.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> indexScrub(@Nonnull SubspaceProvider subspaceProvider, @Nonnull Subspace subspace) {
+
+        final List<Object> additionalLogMessageKeyValues = Arrays.asList(LogMessageKeys.CALLING_METHOD, "indexScrub");
+
+        return iterateAllRanges(additionalLogMessageKeyValues,
+                this::indexScrubRangeOnly,
+                subspaceProvider, subspace);
+    }
+
+    @Nonnull
+    private CompletableFuture<Boolean> indexScrubRangeOnly(@Nonnull FDBRecordStore store, @Nonnull AtomicLong recordsScanned) {
+        // return false when done
+        Index index = common.getIndex();
+        final RecordMetaData metaData = store.getRecordMetaData();
+        final RecordMetaDataProvider recordMetaDataProvider = common.getRecordStoreBuilder().getMetaDataProvider();
+        if (recordMetaDataProvider == null || !metaData.equals(recordMetaDataProvider.getRecordMetaData())) {
+            throw new MetaDataException("Store does not have the same metadata");
+        }
+        final IndexMaintainer maintainer = store.getIndexMaintainer(index);
+        final IndexScrubbingTools<?> tools = maintainer.getIndexScrubbingTools(scrubbingType);
+        if (tools == null) {
+            throw new UnsupportedOperationException("This index does not support scrubbing type " + scrubbingType);
+        }
+
+        return indexScrubRangeOnly(store, recordsScanned, index, tools, maintainer.isIdempotent());
+    }
+
+    private <T> CompletableFuture<Boolean> indexScrubRangeOnly(final @Nonnull FDBRecordStore store, final @Nonnull AtomicLong recordsScanned, final Index index, final IndexScrubbingTools<T> tools, boolean isIdempotent) {
+        // scrubbing only readable
+        validateOrThrowEx(store.getIndexState(index) == IndexState.READABLE, "scrubbed index is not readable");
+
+        final IndexingRangeSet rangeSet = IndexingRangeSet.forScrubbing(scrubbingType, store, index);
+        tools.presetCommonParams(index, scrubbingPolicy.allowRepair(), common.getIndexContext().isSynthetic, common.getAllRecordTypes());
+
+        return rangeSet.firstMissingRangeAsync().thenCompose(range -> {
+            if (range == null) {
+                // Here: no more missing ranges - all done
+                // To avoid stale metadata, we'll keep the scrubbed-ranges indicator empty until the next scrub call.
+                rangeSet.clear();
+                return AsyncUtil.READY_FALSE;
+            }
+            final Tuple rangeStart = RangeSet.isFirstKey(range.begin) ? null : Tuple.fromBytes(range.begin);
+            final Tuple rangeEnd = RangeSet.isFinalKey(range.end) ? null : Tuple.fromBytes(range.end);
+            final TupleRange tupleRange = TupleRange.between(rangeStart, rangeEnd);
+
+            final RecordCursor<T> cursor = tools.getIterator(tupleRange, store, getLimit() + 1); // always respect limit in this path; +1 allows a continuation item in forward scan
```

Comment:
I don't think so. 
The way it used to work was by iterating the cursor until exhausted, then make a second 1 item iteration just to read the next value "after" the cursor. This "+1" is merely a way to do it in a single iteration while respecting the caller's limit. So the cursor returns one extra item - not to be handled but to be used for continuation only. 

I suppose that we could have kept the last handled item as `cont`, and later treat the lowPoint as exclusive.. But doing it without breaking compatibility with ongoing indexing processes might be a challenge. (It could also be a little inefficient when more items are added in-between the last handled and the "+1" unhandled items prior to the next iteration)